File: code/12-JavaCodeLoop/JavaTDDLoop/src/main/java/com/example/valueobjects/Interval.java
package com.example.valueobjects;

/**
 * An inclusive range of integers between {@code lower} and {@code upper}.
 */
public record Interval(int lower, int upper) {

    public static Interval of(int lower, int upper) {
        return new Interval(lower, upper);
    }

    public Interval {
        if (lower > upper) {
            throw new IllegalArgumentException(
                    "lower (" + lower + ") must not be greater than upper (" + upper + ")");
        }
    }
}

File: code/13-Java2Python/python/valueobjects/interval.py
from dataclasses import dataclass


@dataclass(frozen=True)
class Interval:
    """An inclusive range of integers between lower and upper."""

    lower: int
    upper: int

    def __post_init__(self) -> None:
        if self.lower > self.upper:
            raise ValueError(
                f"lower ({self.lower}) must not be greater than upper ({self.upper})"
            )

    @staticmethod
    def of(lower: int, upper: int) -> "Interval":
        return Interval(lower, upper)
